### PR TITLE
feat(#797, #878): set `baseURL` via environment variables and improve internal url detection

### LINKED ISSUE #797: RFC: Improve the handing of `baseURL` to be more intuitive
### Describe the feature

The current logic used to determine if the `auth.baseURL` is internal or external is misleading and hard to understand. The main issue is that the module needs external URLs to be configured in a very specific way to properly identify it as an external URL. 

Examples taken from #782 by @coreyshuman:

```js
auth: {
  baseURL: 'http://base-url',
  provider: {
    endpoints: {
      signIn: {path: 'http://api-url/login'},
    },
}
```

```js
auth: {
  baseURL: 'http://base-url/',
  provider: {
    endpoints: {
      signIn: {path: 'http://api-url/login'},
    },
}
```

```js
auth: {
  baseURL: '',
  provider: {
    endpoints: {
      signIn: {path: 'http://api-url/login'},
    },
}
```

```js
auth: {
  baseURL: 'http://api-url/',
  provider: {
    endpoints: {
      signIn: {path: 'login'},
    },
}
```

In #796 we updated the docs to better explain this behaviour, however we should look into improving this, to make it less complex. 

### How would you implement this?

Some potential ways to better determine if a URL is internal or external:

- Using `schema` (`https` or `http`) to signify a URL is external
- Ignoring trailing and leading `/` inside the `baseURL` and `endpoints` and then composing and adding them as needed, to better infer the URLs, instead of just combining them

More suggestions are welcome!

### Additional information

- [ ] Would you be willing to help implement this feature?

### Provider

- [ ] AuthJS
- [ ] Local
- [ ] Refresh
- [ ] New Provider

### PR DESCRIPTION
**Pull request by zoey-kaiser**
### 🔗 Linked issue

closes #797, #878

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation, readme or JSdoc annotations)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [X] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [x] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📝 Checklist

- [ ] I have linked an issue or discussion.
- [ ] I have added tests (if possible).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by zoey-kaiser** [comment]
## Functional tests:

### Only setting `auth.baseURL`

- `auth.baseURL` not set: 
  - Requests made to `/api/auth`
- `auth.baseURL` set to `http://localhost:8080`
  - Requests made to `http://localhost:8080/api/auth`
- `auth.baseURL` set to `http://localhost:8080/auth`
  - Requests made to `http://localhost:8080/auth`

### Combination of `auth.baseURL` and `AUTH_ORIGIN`

- `auth.baseURL` not set, `AUTH_ORIGIN` not set:
  - Requests made to `/api/auth`
- `auth.baseURL` set to `http://localhost:8080`, `AUTH_ORIGIN` not set:
  - Requests made to `http://localhost:8080/api/auth`
- `auth.baseURL` set to `http://localhost:8080`, `AUTH_ORIGIN` set to `http://localhost:8081`:
  - Requests made to `http://localhost:8081/api/auth`
- `auth.baseURL` set to `http://localhost:8080`, `AUTH_ORIGIN` set to `http://localhost:8081/auth`:
  - Requests made to `http://localhost:8081/auth`

File: playground-authjs/nuxt.config.ts [hunk 1]
```diff
@@ -8,7 +8,7 @@ export default defineNuxtConfig({
     globalAppMiddleware: {
       isEnabled: true
     },
-    baseURL: `http://localhost:${process.env.PORT || 3000}`
+    baseURL: `http://localhost:${process.env.PORT || 3000}/api/auth`
```
Comment (phoenix-ru) [inline playground-authjs/nuxt.config.ts:11]: This looks like a breaking change

**Review by phoenix-ru — COMMENTED** [review]
In general I understand the intent and very like the testing comment you left. However, there are several points:

1. What happens to users who already use `baseURL` as e.g. `http://localhost:3000`? In my understanding, requests are now made to this URL without adding `/api/auth`? This is a breaking change and needs to have its own minor release + docs if we settle on it.
2. What happens to internal `$fetch` calls? Remember that they have a limitation - they only work when path starts with a `/`. And here are two facts that break it:
  1. When using `authjs` provider in which these calls are the most relevant, you need to set `AUTH_ORIGIN` or `baseURL` to a fully-specified URL (i.e. protocol, hostname, etc.);
  2. Calling a fully-specified URL using `$fetch` invokes _external_ fetch, meaning that a real HTTP call is made from the server to itself - and it has a relatively high cost, bottlenecking performance.

Any implementation that doesn't provide backwards compatibility with at least these two points is therefore highly discouraged :/
I'd like for us to get involved in an RFC discussion and come up with a spec of how calls are made depending on what variable is set. For example, we know for sure:
- `authjs` provider is the same Nuxt server and therefore should always prefer internal calls, regardless of the `AUTH_ORIGIN` (only taking into account `pathname`);
- `local` provider is isomorphic (with lean towards external backends, it seems), and origin needs to be taken literally - if it is set, we assume all calls are external.
With these in mind, it might even make sense to disconnect the implementations of two providers into separate URL utils.

**Review by phoenix-ru — APPROVED** [review]

**Comment by shayr1** [comment]
I am using http://localhost:8000/api and then 
```
endpoints: {
        signIn: { path: '/auth/login', method: 'post' },
        signOut: { path: '/auth/logout', method: 'post' },
        signUp: { path: '/auth/register', method: 'post' },
        getSession: { path: '/auth/user', method: 'get' },
      }
```
I am not using `AUTH_ORIGIN`, just according to instructions above, however it is always using `http://localhost:3000`

I am using v0.10.0

**Pull request merged**